### PR TITLE
useOpenAIStream: Improve error handling

### LINKED ISSUE #76211: Hardening: Cover edge cases (network goes down, OpenAI goes down / token rate limit exceeded, etc)
Hardening work for `useOpenAIStream` around error handling:
- The component doesn't unsubscribe from the observer that manages the stream when unmounting
- The connection is not closed after X seconds waiting for the stream to start receiving data
- The error is too complex to understand. Display a generic error and the full error in the console

### PR DESCRIPTION
**Pull request by ivanortegaalba**
**What does this PR do?**
- Simplify the error when the OpenAI call fails. The full error is displayed now in the console
- Unsubscribe from the Observable when the component unmounts
- Reset the state and time out when the stream is not generating data for 10s

**Why do we need this feature?**

There are cases where the backend doesn't error, but it doesn't get any data when generating. From the frontend, we want to time out and close the stream connection.

**Which issue(s) does this PR fix?**:

Fixes #76211

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/dashboard/components/GenAI/GenAIButton.test.tsx [hunk 1]
```diff
@@ -208,7 +208,9 @@ describe('GenAIButton', () => {
       // The tooltip keeps interactive to be able to click the link
       await userEvent.hover(tooltip);
       expect(tooltip).toBeVisible();
-      expect(tooltip).toHaveTextContent('Something went wrong');
+      expect(tooltip).toHaveTextContent(
+        'Failed to generate content using OpenAI. Please try again or if the problem persist, contact your organization admin.'
```
Comment (nmarrs) [inline public/app/features/dashboard/components/GenAI/GenAIButton.test.tsx:212]:
```suggestion
        'Failed to generate content using OpenAI. Please try again or if the problem persists, contact your organization's admin.'
```

**Review by nmarrs — APPROVED** [review]
LGTM - just a small wording nit :)

File: public/app/features/dashboard/components/GenAI/hooks.ts [hunk 2]
```diff
@@ -46,6 +48,21 @@ export function useOpenAIStream(
...
   const { error: notifyError } = useAppNotification();
 
+  const onError = useCallback(
+    (e: Error) => {
+      setStreamStatus(StreamStatus.IDLE);
+      setMessages([]);
+      setError(e);
+      notifyError(
+        'Failed to generate content using OpenAI',
+        `Please try again or if the problem persist, contact your organization admin.`
```
Comment (ivanortegaalba) [inline public/app/features/dashboard/components/GenAI/hooks.ts:58]:
```suggestion
        `Please try again or if the problem persists, contact your organization admin.`
```

File: public/app/features/dashboard/components/GenAI/hooks.ts [hunk 3]
```diff
@@ -99,6 +110,28 @@ export function useOpenAIStream(
...
+  // Unsubscribe from the stream when the component unmounts.
+  useEffect(() => {
+    return () => {
+      if (value?.stream) {
+        value.stream.unsubscribe();
+      }
+    };
+  }, [value]);
+
+  // If the stream is generating and we haven't received a reply, it timed out.
```
Comment (ivanortegaalba) [inline public/app/features/dashboard/components/GenAI/hooks.ts:122]:
```suggestion
  // If the stream is generating and we haven't received a reply, it times out.
```

**Review by adela-almasan — APPROVED** [review]
✨

File: public/app/features/dashboard/components/GenAI/GenAIButton.test.tsx (hunk 1 above)
Comment (imatwawana) [inline public/app/features/dashboard/components/GenAI/GenAIButton.test.tsx:212]: This should read:
'Failed to generate content using OpenAI. Please try again, or if the problem persists, contact your organization admin.'

File: public/app/features/dashboard/components/GenAI/GenAIButton.tsx [hunk 4]
```diff
@@ -173,7 +173,13 @@ export const GenAIButton = ({
     <div className={styles.wrapper}>
       {isFirstHistoryEntry && <Spinner size={14} />}
       {!hasHistory && (
-        <Tooltip show={error ? undefined : false} interactive content={`OpenAI error: ${error?.message}`}>
+        <Tooltip
+          show={error ? undefined : false}
+          interactive
+          content={
+            'Failed to generate content using OpenAI. Please try again or if the problem persist, contact your organization admin.'
```
Comment (imatwawana) [inline public/app/features/dashboard/components/GenAI/GenAIButton.tsx:180]: This should read:
'Failed to generate content using OpenAI. Please try again, or if the problem persists, contact your organization admin.'

File: public/app/features/dashboard/components/GenAI/hooks.ts [hunk 5]
```diff
@@ -46,6 +48,21 @@ export function useOpenAIStream(
...
   const { error: notifyError } = useAppNotification();
 
+  const onError = useCallback(
+    (e: Error) => {
+      setStreamStatus(StreamStatus.IDLE);
+      setMessages([]);
+      setError(e);
+      notifyError(
+        'Failed to generate content using OpenAI',
+        `Please try again or if the problem persists, contact your organization admin.`
```
Comment (imatwawana) [inline public/app/features/dashboard/components/GenAI/hooks.ts:58]: This should read (just adding a comma):
`Please try again, or if the problem persists, contact your organization admin.`

**Review by imatwawana — COMMENTED** [review]
Added the spots where there a comma and an "s" is needed.

**Pull request merged**